### PR TITLE
Extract drive-by fixes from PR 12135 for easier reviewing

### PR DESCRIPTION
**Pull request by itsjunetime**
## Which issue does this PR close?

Helps with reviewing of #12135 

## Rationale for this change

These are just a few changes that make the code cleaner/easier to understand/etc that I added when filing #12135. It was pointed out to me, though, that these make the PR harder to review as there's a lot of changes not directly related to the PR. So I'm pulling those changes into this PR so that they are hidden from the diff once this is merged, thus making the PR easier to review.

## What changes are included in this PR?

A few clean-ups that shouldn't change behavior at all.

## Are these changes tested?

All tests pass just like they do on main.

## Are there any user-facing changes?

Shouldn't be

### CODE REVIEW

**Comment by itsjunetime** [comment]
It looks like clippy failed due to some internal CI connectivity issue - I don't think I can rerun them by myself.

**Comment by alamb** [comment]
> It looks like clippy failed due to some internal CI connectivity issue - I don't think I can rerun them by myself.

Thanks @itsjunetime 

I have restarted the CI (you do need to be a datafusion committer to restart the jobs individually)

One trick to get github to rerun the CI is to close the PR and then reopen it (that will trigger rerunning the whole PR)

File: datafusion/core/src/datasource/listing/helpers.rs [hunk 1]
```diff
@@ -51,12 +51,12 @@ use object_store::{ObjectMeta, ObjectStore};
 /// - the table provider can filter the table partition values with this expression
 /// - the expression can be marked as `TableProviderFilterPushDown::Exact` once this filtering
 ///   was performed
-pub fn expr_applicable_for_cols(col_names: &[String], expr: &Expr) -> bool {
+pub fn expr_applicable_for_cols(col_names: &[&str], expr: &Expr) -> bool {
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/helpers.rs:54]: This is a nice removal of some clones 👍

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 2]
```diff
@@ -133,12 +134,15 @@ impl std::fmt::Display for BinaryExpr {
 }
 
 /// Invoke a boolean kernel on a pair of arrays
-macro_rules! boolean_op {
-    ($LEFT:expr, $RIGHT:expr, $OP:ident) => {{
-        let ll = as_boolean_array($LEFT).expect("boolean_op failed to downcast array");
-        let rr = as_boolean_array($RIGHT).expect("boolean_op failed to downcast array");
-        Ok(Arc::new($OP(&ll, &rr)?))
-    }};
+#[inline]
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:137]: 👍

**Review by alamb — APPROVED** [review]
Looks like a very nice set of cleanups to me @itsjunetime -- thank you

**Pull request merged**